testing/integration/fqdn_test.go

```diff
@@ -0,0 +1,292 @@
+	s.verifyHostNameInIndices("logs-*", shortName)
+	s.verifyHostNameInIndices("metrics-*", shortName)
+
+	// Update Agent policy to enable FQDN
+	policy.AgentFeatures = []map[string]interface{}{
+		{
+			"name":    "fqdn",
+			"enabled": true,
+		},
+	}
+	updatePolicyReq := kibana.UpdatePolicyRequest(*policy)
+	_, err = kibClient.UpdatePolicy(updatePolicyReq)
+	require.NoError(s.T(), err)
+
+	// Wait until policy has been applied by Agent
+	prevAgentPolicyRevision := agent.PolicyRevision
+	require.Eventually(s.T(), func() bool {
+		getAgentReq := kibana.GetAgentRequest{ID: agent.ID}
+		updatedPolicyAgent, err := kibClient.GetAgent(getAgentReq)
+		require.NoError(s.T(), err)
+
+		if updatedPolicyAgent.PolicyRevision > prevAgentPolicyRevision {
+			prevAgentPolicyRevision = updatedPolicyAgent.PolicyRevision
+			return true
+		}
+
+		return false
+	}, 30*time.Second, 1*time.Second)
+
+	// Verify that agent name is FQDN
+	agent, err = tools.GetAgentByHostnameFromList(kibClient, fqdn)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), agent)
+
+	// Verify that hostname in `logs-*` and `metrics-*` is FQDN
+	s.verifyHostNameInIndices("logs-*", fqdn)
+	s.verifyHostNameInIndices("metrics-*", fqdn)
+
+	// Update Agent policy to disable FQDN
+	policy.AgentFeatures = []map[string]interface{}{
+		{
+			"name":    "fqdn",
+			"enabled": false,
+		},
+	}
+	updatePolicyReq = kibana.UpdatePolicyRequest(*policy)
+	_, err = kibClient.UpdatePolicy(updatePolicyReq)
+	require.NoError(s.T(), err)
+
+	// Wait until policy has been applied by Agent
+	require.Eventually(s.T(), func() bool {
+		getAgentReq := kibana.GetAgentRequest{ID: agent.ID}
+		updatedPolicyAgent, err := kibClient.GetAgent(getAgentReq)
+		require.NoError(s.T(), err)
+
+		return updatedPolicyAgent.PolicyRevision > prevAgentPolicyRevision
+	}, 30*time.Second, 1*time.Second)
+
+	// Verify that agent name is short hostname again
+	agent, err = tools.GetAgentByHostnameFromList(kibClient, shortName)
+	require.NoError(s.T(), err)
+	require.NotNil(s.T(), agent)
+
+	// Verify that hostname in `logs-*` and `metrics-*` is short hostname again
+	s.verifyHostNameInIndices("logs-*", shortName)
+	s.verifyHostNameInIndices("metrics-*", shortName)
+}
+
+func (s *FQDN) verifyHostNameInIndices(indices, hostname string) {
+	search := s.requirementsInfo.ESClient.Search
+	resp, err := search(
+		search.WithIndex(indices),
+		search.WithSort("@timestamp:desc"),
+		search.WithFilterPath("hits.hits"),
+	)
+	require.NoError(s.T(), err)
+	require.False(s.T(), resp.IsError())
+	defer resp.Body.Close()
+
+	var body struct {
+		Hits struct {
+			Hits []struct {
+				Source struct {
+					Host struct {
+						Name string `json:"name"`
+					} `json:"host"`
+				} `json:"_source"`
+			} `json:"hits"`
+		} `json:"hits"`
+	}
+	decoder := json.NewDecoder(resp.Body)
+	err = decoder.Decode(&body)
+	require.NoError(s.T(), err)
+
+	for _, hit := range body.Hits.Hits {
+		require.Equal(s.T(), hostname, hit.Source.Host.Name)
```

Comment:
[Suggestion]
I'd not use `require`, it's a case where it's worth moving on if this assertion fails. That way, we can stop the test after both indices are checked and know which ones are problematic. Or even just keep going and at the end of the test see all that failed.

If the test stops at the first failure, it might get fixed, just to have the test breaking at the next check. Knowing all the failures at the end allows for a quick feedback loop.

Reply:
++ to not using `require` inside loops like the one here.  And also agree about not using `require` in long-running tests like this E2E test.  The one exceptions I see are the use of `require.NoErr` and maybe `require.NotNil`, mostly because these checks also usually act like guards against NPEs down the line.